lib/secrets/secrets.go

```diff
@@ -229,3 +299,127 @@ func PrepareSSHPrivateKey(keyContent string) (string, error) {
+// returns the ciphertext, iv and error
+func encrypt(key []byte, value []byte) ([]byte, []byte, error) {
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "error creating cipher block during encryption")
+	}
+	iv, err := GenerateRandomBytes(block.BlockSize())
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "could not generate random iv for encryption")
+	}
+	b := hex.EncodeToString(value)
+	stream := cipher.NewCTR(block, iv)
+	ciphertext := make([]byte, len(b))
+	stream.XORKeyStream(ciphertext, []byte(b))
+	return ciphertext, iv, nil
+}
+
+// decrpting the cipher for ctr mode
+// returns the decoded data ( originally encoded with base64 ) and error
+func decrypt(block cipher.Block, ciphertext []byte, iv []byte) ([]byte, error) {
+	stream := cipher.NewCTR(block, iv)
+	plain := make([]byte, len(ciphertext))
+	stream.XORKeyStream(plain, ciphertext)
+	d, err := hex.DecodeString(string(plain))
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to decode secret value while decryption")
+	}
+	return d, nil
+}
+
+// function used to write the secret values to given file in toml format
+// also sets the owner for that file
+// stores the name of the algorithm , iv and the ciphertext
+func writeToml(algorithm string, iv string, ciphertext string, path string, ownerUID int, ownerGID int) error {
+	secretData := SecretKeyToml{
+		Algorithm:  algorithm,
+		IV:         iv,
+		Ciphertext: ciphertext,
+	}
+
+	buf := new(bytes.Buffer)
+	if err := toml.NewEncoder(buf).Encode(secretData); err != nil {
+		return errors.Wrap(err, "failed to encode struct SecretKeyToml data to toml")
+	}
+
+	err := fileutils.AtomicWrite(path, bytes.NewReader(buf.Bytes()), fileutils.WithAtomicWriteFileMode(0700))
+	if err != nil {
+		return errors.Wrap(err, "error writing the secret file")
+	}
+
+	// TODO(ssd) 2018-08-20: Should this be an option we can pass to AtomicWrite?
+	err = os.Chown(path, ownerUID, ownerGID)
+	if err != nil {
+		return errors.Wrap(err, "could not chown secrets data")
+	}
+	return nil
+}
+
+// takes in the secret value and encrypts it
+// return encrypted cipher, iv and error
+func getEncryptedData(data []byte, basePath string) ([]byte, []byte, error) {
+	// we create the encryption key if it doesnot exist
+	var secretKey []byte
+	keyExists, err := keyExists(basePath)
+	if err != nil {
+		return nil, nil, err
+	}
+	if !keyExists {
+		return nil, nil, errors.New("Could not find the encryption key")
+	}
+	path := filepath.Join(basePath, "key")
+	secretKey, err = ioutil.ReadFile(path)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "could not read secrets key (unexpected error)")
+	}
+	encryptedData, iv, err := encrypt(secretKey, data)
+	if err != nil {
+		return nil, nil, err
+	}
+	return encryptedData, iv, nil
+}
+
+// gets the secret from the toml file and decrypts it
+// returns the decrypted data or error
+func getDecryptedData(basePath string, ret []byte) ([]byte, error) {
```

Comment:
I think the name ret doesn't make sense here as its not the returned value

Reply:
changed it to data